hub/core/io.py

```diff
@@ -0,0 +1,363 @@
+from hub.core.meta.encode.chunk_id import CHUNK_ID_COLUMN, ChunkIdEncoder
+from hub.core.storage import LRUCache, MemoryProvider, StorageProvider, LocalProvider
+from hub.util.exceptions import (
+    DatasetUnsupportedPytorch,
+    SampleDecompressionError,
+    TensorDoesNotExistError,
+)
+from hub.util.keys import get_chunk_key
+from hub.util.remove_cache import get_base_storage
+from hub.util.storage import get_pytorch_local_storage
+
+ChunkEngineMap = Dict[str, ChunkEngine]
+CachesMap = Dict[str, LRUCache]
+
+
+class IOBlock:
+    """
+    Represents ordered sequential read of samples from corresponding tensor chunks.
+    """
+
+    def __init__(self, chunks: List[str], indexes: List[int]) -> None:
+        self._chunks: List[str] = chunks
+        self._ind: List[int] = indexes
+
+    def shuffle(self):
+        r"""
+        Shuffle sequence in which indices would be read from the IOBlock
+        """
+        shuffle(self._ind)
+
+    def chunk_name(self, tensor_index: int) -> str:
+        return self._chunks[tensor_index]
+
+    def indices(self) -> List[int]:
+        return self._ind
+
+    def chunks(self) -> List[str]:
+        return self._chunks
+
+    def __len__(self) -> int:
+        return len(self._ind)
+
+
+class Schedule:
+    def __init__(self, blocks: List[IOBlock]) -> None:
+        self._blocks: List[IOBlock] = blocks
+
+    def shuffle(self) -> None:
+        r"""
+        Shuffle IOBlocks in the schedule as well as each IOBlock
+        """
+        shuffle(self._blocks)
+
+        for block in self._blocks:
+            block.shuffle()
+
+    def __iter__(self):
+        return iter(self._blocks)
+
+    def __len__(self):
+        return sum(map(len, self._blocks))
+
+
+class Scheduler(ABC):
+    @abstractmethod
+    def schedule(self, jobs: List[IOBlock]) -> List[Schedule]:
+        ...
+
+
+class SingleThreadScheduler(Scheduler):
+    def schedule(self, jobs: List[IOBlock]) -> List[Schedule]:
+        return [Schedule(jobs)]
+
+
+class SequentialMultithreadScheduler(Scheduler):
+    """
+    Splits list of IO blocks in a way, so PyTroch loader would return
+    samples in sequence, when started with `num_worker` > 1.
+
+    Scheduler relays on a fact, that PyTroch DataLoader synchronize
+    read of samples per thread and return in a sequence per worker.
+
+    Example:
+        Given sequence of indices `[1, 2, 3, 4, 5, 6]` and 4 workers
+        PyTorch have to read samples in an order
+            thread0: [1, 5]
+            thread1: [2, 6]
+            thread2: [3]
+            thread3: [4]
+
+        So that initial order would be reconstructed by the DataLoader
+    """
+
+    def __init__(self, num_workers: int) -> None:
+        super().__init__()
+        self.num_workers = num_workers
+
+    def schedule(self, jobs: List[IOBlock]) -> List[Schedule]:
+        per_worker: List[List[IOBlock]] = [list() for _ in range(self.num_workers)]
+        assigneg_worker = iter(cycle(range(self.num_workers)))
```

Comment:
did you mean `assigned_worker`? A docstring for this method would help, it's difficult to parse at first glance.

Reply:
Fixed